File: distributed_median/distributed_median.py
"""
Suppose there are N machines, each with a list of random integers.
Find a way to calculate the median across the different machines.
"""

import random

# ---------------------------------------------------------
def find_element_k(nums, k):
    while True:
        # Remove empty lists
        nums = [i for i in nums if len(i) != 0]

        # Find number of element less than a random pivot element
        machine_i = random.randint(0, len(nums)-1)
        element_i = random.randint(0, len(nums[machine_i])-1)
        element = nums[machine_i][element_i]
        num_lte = [len([j for j in i if j <= element]) for i in nums]
        num_lte = sum(num_lte)
        num_e = [len([j for j in i if j == element]) for i in nums]
        num_e = sum(num_e)
        num_elements = [len(i) for i in nums]
        num_elements = sum(num_elements)
        # print([nums, element, num_lte, num_e, num_elements])

        # Current pivot element is the kth element
        if num_lte == k or num_e == num_elements:
            return element
        # Current pivot is too small
        # The number of less-than elements we are expecting should also decrease
        elif num_lte <= k:
            nums = [[j for j in i if j > element] for i in nums]
            k -= num_lte
        # Current pivot element is too large
        # Removing elements larger than the pivot does not affect num_lte
        elif num_lte > k:
            nums = [[j for j in i if j <= element] for i in nums]

def distributed_median(nums):
    # Out of 7 elements, median is element #4
    num_elements = [len(i) for i in nums]
    num_elements = sum(num_elements)
    if num_elements % 2 != 0:
        i = find_element_k(nums, num_elements//2+1)
        return i
    else:
        i1 = find_element_k(nums, num_elements//2+1)
        i2 = find_element_k(nums, num_elements//2)
        return (i1+i2)/2

def exact_median(nums):
    flattened = []
    for i in nums:
        for j in i:
            flattened.append(j)
    flattened = sorted(flattened)
    if len(flattened) % 2 != 0:
        i = flattened[len(flattened)//2]
        return i
    else:
        i1 = flattened[len(flattened)//2]
        i2 = flattened[len(flattened)//2 - 1]
        return (i1+i2)/2

# ---------------------------------------------------------
nums = [[1,2,3], [4], [5,6,7]]
result = exact_median(nums)
print("Exact median: {}".format(result))
result = distributed_median(nums)
print("Distributed median: {}".format(result))
print()

nums = [[1,2,3], [4], [5,6,7,8]]
result = exact_median(nums)
print("Exact median: {}".format(result))
result = distributed_median(nums)
print("Distributed median: {}".format(result))
print()

nums = [[1], [1]]
result = exact_median(nums)
print("Exact median: {}".format(result))
result = distributed_median(nums)
print("Distributed median: {}".format(result))
print()

num_machines = random.randint(10, 20)
num_elements = random.choices(range(10), k=num_machines)
nums = [[random.randint(0, 1000) for j in range(i)] for i in num_elements]
result = exact_median(nums)
print("Exact median: {}".format(result))
result = distributed_median(nums)
print("Distributed median: {}".format(result))
print()

File: distributed_median/distributed_median.cpp
// g++ -std=c++17 distributed_median.cpp
/*
Suppose there are N machines, each with a list of random integers.
Find a way to calculate the median across the different machines.
*/

#include <iostream>    // std::cout, std::endl
#include <random>      // std::random_device, std::mt19937,
                       // std::uniform_int_distribution
#include <vector>      // std::vector

std::random_device rd;

// ------------------------------------------------------------------
template<typename T>
void print_nums(T& v) {
    std::cout << "[";
    for (int i = 0; i < v.size(); i++) {
        std::cout << "[";
        for (int j = 0; j < v[i].size(); ++j) {
            if (j == v[i].size() - 1) {
                std::cout << v[i][j];
            } else {
                std::cout << v[i][j] << ", ";
            }
        }
        if (i == v.size() - 1) {
            std::cout << "]";
        } else {
            std::cout << "], ";
        }
    }
    std::cout << "]" << std::endl;
}

// ------------------------------------------------------------------
double find_element_k(std::vector<std::vector<int>> nums, int k) {
    while (true) {
        // Remove empty lists
        for (std::vector<std::vector<int>>::iterator it = nums.begin();
             it != std::end(nums); ) {
            if (it->size() == 0) {
                it = nums.erase(it);
            } else {
                ++it;
            }
        }

        // Randomly select a pivot element
        int machine_i = std::uniform_int_distribution<int>(0, nums.size()-1)(rd);
        int element_i = std::uniform_int_distribution<int>(0, nums[machine_i].size()-1)(rd);
        double element = nums[machine_i][element_i];

        // Find number of element less than a random pivot element
        int num_lte = 0;
        int num_e = 0;
        int num_elements = 0;
        for (auto& v : nums) {
            for (auto& i : v) {
                if (i <= element) {
                    num_lte += 1;
                }
                if (i == element) {
                    num_e += 1;
                }
                num_elements += 1;
            }
        }
        // print_nums(nums);
        // std::cout << element << ", ";
        // std::cout << num_lte << ", ";
        // std::cout << k << ", ";
        // std::cout << num_e << ", ";
        // std::cout << num_elements << std::endl;

        // Current pivot element is the kth element
        if (num_lte == k || num_e == num_elements) {
            return element;
        }
        // Current pivot is too small
        // The number of less-than elements we are expecting should also decrease
        else if (num_lte <= k) {
            for (auto& v : nums) {
                for (std::vector<int>::iterator it = v.begin();
                     it != std::end(v); ) {
                    if (*it <= element) {
                        it = v.erase(it);
                    } else {
                        ++it;
                    }
                }
            }
            k -= num_lte;
        }
        // Current pivot element is too large
        // Removing elements larger than the pivot does not affect num_lte
        else if (num_lte > k) {
            for (auto& v : nums) {
                for (std::vector<int>::iterator it = v.begin();
                     it != std::end(v); ) {
                    if (*it > element) {
                        it = v.erase(it);
                    } else {
                        ++it;
                    }
                }
            }
        }
    }
}

double distributed_median(std::vector<std::vector<int>> nums) {
    // Calculate the number of elements
    int num_elements = 0;
    for (auto& v : nums) {
        num_elements += v.size();
    }

    // Calculate median
    if (num_elements % 2 != 0) {
        double i = find_element_k(nums, num_elements/2 + 1);
        return i;
    } else {
        double i1 = find_element_k(nums, num_elements/2 + 1);
        double i2 = find_element_k(nums, num_elements/2);
        return (i1+i2)/2;
    }
}

double exact_median(std::vector<std::vector<int>> nums) {
    // Flatten & sort
    std::vector<int> flattened;
    for (auto& v : nums) {
        for (auto& i : v) {
            flattened.push_back(i);
        }
    }
    std::sort(flattened.begin(), flattened.end());

    // Calculate median
    if (flattened.size() % 2 != 0) {
        double i = flattened[flattened.size()/2];
        return i;
    } else {
        double i1 = flattened[flattened.size()/2];
        double i2 = flattened[flattened.size()/2 - 1];
        return (i1+i2)/2;
    }
}

// ------------------------------------------------------------------
int main() {
    // Set up
    std::vector<std::vector<int>> nums;
    double result;

    // Answer: 4
    nums = {{4}, {1,2,3}, {5,6,7}};
    result = exact_median(nums);
    std::cout << "Exact median: " << result << std::endl;
    result = distributed_median(nums);
    std::cout << "Distributed median: " << result << std::endl;
    std::cout << std::endl;

    // Answer: 4.5
    nums = {{1,2,3}, {4}, {5,6,7,8}};
    result = exact_median(nums);
    std::cout << "Exact median: " << result << std::endl;
    result = distributed_median(nums);
    std::cout << "Distributed median: " << result << std::endl;
    std::cout << std::endl;

    // Answer: 1.0
    nums = {{1}, {1}};
    result = exact_median(nums);
    std::cout << "Exact median: " << result << std::endl;
    result = distributed_median(nums);
    std::cout << "Distributed median: " << result << std::endl;
    std::cout << std::endl;
}